mypy/binder.py

```diff
@@ -0,0 +1,288 @@
+from contextlib import contextmanager
+from typing import (Any, Dict, List, cast, Iterator)
+
+from mypy.types import Type, AnyType, PartialType
+from mypy.nodes import (Node, Var)
+
+from mypy.subtypes import is_subtype
+from mypy.join import join_simple, join_types
+from mypy.sametypes import is_same_type
+
+
+class Frame(Dict[Any, Type]):
+    """Frame for the ConditionalTypeBinder.
+
+    Frames actually in the binder must have
+    binder.frames[i].parent == binder.frames[i-1].
+
+    options_on_return represents a list of possible frames that we
+    will incorporate next time we are the top frame (e.g. frames from
+    the end of each if/elif/else block).
+
+    After a frame is complete, 'changed' represents whether anything
+    changed during the procedure, so a loop should be re-run.
+    'breaking' represents whether frame was necessarily breaking out.
+    """
+
+    changed = False
+    broken = False
+
+    def __init__(self, parent: 'Frame' = None) -> None:
+        super().__init__()
+        self.parent = parent
+        self.options_on_return = []  # type: List[Frame]
+
+    def add_return_option(self, frame: 'Frame') -> None:
+        """When this frame is next visited, it may may have state `frame`.
```

Comment:
Double "may". Also maybe it's clearer to say something like: The frame `frame` (which must be a descendent of this one) may jump to the current point in this frame.